Rework and fixes

WarningImHack3r opened the pull request:
- Fix README lint
- Update to `manifest.json` v3
- Remove useless `plex.tv` domain from manifest (_this should only be working on app.plex.tv shouldn't it?_)
- Fix little stuff on injector
- Fix download button not working (update regex declaration, fix xml load handling, add injection on first page load, add more relevant error alerts, fix code here and there)
- Move global variables to inline

Future improvement could be allowing bulk downloads for TV shows e.g.

WarningImHack3r committed:
Rework and fixes
- Fix README lint
- Update to manifest.json v3
- Remove useless plex.tv domain from manifest (this should
only be working on app.plex.tv shouldn't it?)
- Fix little stuff on injector
- Fix download button not working (update regex declaration, fix xml load
handling, add more relevant error alerts, fix code here and there, add injection
on first page load)

WarningImHack3r committed:
More fixes
- Move global variables to inline
- Improve '?download=1' handling
- Reduce timeout for first load

WarningImHack3r committed:
Forgot download=1

jkingsman commented on src/js/download_script.js:
```diff
@@ -13,68 +13,58 @@
                 if (request.readyState == 4 && request.status == 200) {
                     callback(request.responseXML);
+                } else {
+                    alert("Unsuccessful XML URL: " + url);
```
I'd rather make these `console.error()`s at most; better to fail silently than be disruptive.

jkingsman commented on src/js/download_script.js:
```diff
@@ -13,68 +13,58 @@
 
         const getMetadata = function (xml) {
-            const clientId = clientIdRegex.exec(window.location.href);
+            const clientId = /server\/([a-f0-9]{40})\//.exec(window.location.href);
```
Could you explain why you inlined these? They seemed easy to edit and update when they were all grouped together 🙂

WarningImHack3r replied:
Because they were used only once so they looked like wasted lines, but you're right I'm gonna revert that

jkingsman commented:
Heyya! Thanks so much for the updates and changes. Looks good overall; not sure I'm a huge fan of inlining the XML paths but it's not the end of the world.

I'll run this through some tests tomorrow when I've got some spare time and then merge + update the store entry.

WarningImHack3r committed:
Improvements
- Fix download button not appearing thanks to a dynamic observer
- Fix download link because of metadataId

WarningImHack3r commented:
I've added some important improvements. I'm not an expert in JavaScript so I don't know if I did them in the best possible way but it works like a charm, much better than with `setTimeout`s.  
Will revert the inline-thing lines back right now and we should be good.

WarningImHack3r commented:
> Will revert the inline-thing lines back right now and we should be good.

While I'm doing that, can we discuss about the issue #5? Removing this flag causes a slight misbehave with the output file name, and the download url with Plex Pass does contain it. I've tested it a lot and I've never encountered such an issue.  
Should I keep it?

WarningImHack3r committed:
Revert inline strings & fix accessToken
Now I understand why I shouldn't have removed it

WarningImHack3r commented:
Starting noticing that I reverted **a lot** of things since my first commit, and basically this PR fixes (mainly) an XML path, improves button creation so that it never fails and updates to manifest v3 haha

WarningImHack3r committed:
Changes in content_script

jkingsman commented:
Hi! Sorry for disappearing; I've had family over and things have been hectic.

> Will revert the inline-thing lines back right now and we should be good.

Awesome; thanks! Just think it helps with readability.

> I'm not an expert in JavaScript so I don't know if I did them in the best possible way but it works like a charm, much better than with setTimeouts.

Yeah mutationobserver is a much nicer solution; thanks for that.

> While I'm doing that, can we discuss about the issue https://github.com/jkingsman/plex-shared-library-downloader/issues/5? Removing this flag causes a slight misbehave with the output file name, and the download url with Plex Pass does contain it. I've tested it a lot and I've never encountered such an issue.
Should I keep it?

Can you elaborate on the misbehaving filename? I agree that n=1 is a low number to change behavior on, but I haven't seen any complaints about the way it currently is. Unless it's egregious, I'm inclined to leave it as is, but I'm interested to hear about the filename issue.

> Starting noticing that I reverted a lot of things since my first commit, and basically this PR fixes (mainly) an XML path, improves button creation so that it never fails and updates to manifest v3 haha

No worries; I appreciate the work! In the future, making sure you match quotation mark types can help reduce diff noise but it's not a huge deal.

I'll wait to hear back from you on the filename issues and make a decision on that.

Thanks again for looking at this! I don't get as much time to spend on side projects as I used to, so the help is welcome.

WarningImHack3r commented:
> Hi! Sorry for disappearing; I've had family over and things have been hectic.

No problem! Hope everything is okay.
> > Will revert the inline-thing lines back right now and we should be good.
> 
> Awesome; thanks! Just think it helps with readability.

Done; I quite like it too
> > While I'm doing that, can we discuss about the issue #5? Removing this flag causes a slight misbehave with the output file name, and the download url with Plex Pass does contain it. I've tested it a lot and I've never encountered such an issue.
> > Should I keep it?
> 
> Can you elaborate on the misbehaving filename? I agree that n=1 is a low number to change behavior on, but I haven't seen any complaints about the way it currently is. Unless it's egregious, I'm inclined to leave it as is, but I'm interested to hear about the filename issue.

The main point is that official download links (when hovering `Download file` in media menu if you have Plex Pass) do contain `download=1`, so why not keeping it.  
Then, based on my observations, when you *don't* put `download=1`, the target file obtained is always named `file.mkv` (or `.avi`/`.mp4`...), and not with the server media name, which might not be the best thing.  
So in my opinion, this parameter *must* stay, and I have no clue why/how someone could have an issue with it.  
(I'm thinking, maybe using a right click on the button to use the without-`download=1` URL could be a good idea? Or something else if you *really* don't want it?)

> Thanks again for looking at this! I don't get as much time to spend on side projects as I used to, so the help is welcome.

No problem! No other "free Plex download" tool was working, and my friends needed one to download from my server. This one seemed the easiest to fix, so I did it.

WarningImHack3r committed:
Add mention of Safari extension

WarningImHack3r commented:
Just made a Safari version!

jkingsman commented:
Hm alright, points taken on the `download=1`. Let's leave it the way that you have it for now; if someone comes back with an issue we can dig in deeper. I agree emulating Plex is the best approach and swapping it out on an n=1 report was perhaps premature.

I think this is good to merge and update in the store; anything else you want to add or should I pull the trigger? Also cool about the Safari extension 😎👌

WarningImHack3r commented:
Nice! No everything looks fine on my side, ready for your merge!

jkingsman commented:
Merged and submitted for review! Thanks again for your help and work; I really appreciate it. Enjoy your weekend!
